applications/FluidDynamicsApplication/custom_elements/fluid_element.cpp

```diff
@@ -0,0 +1,788 @@
+
+
+template< class TElementData >
+void FluidElement<TElementData>::GetSecondDerivativesVector(Vector &rValues, int Step)
+{
+    GeometryType& r_geometry = this->GetGeometry();
+
+    if (rValues.size() != LocalSize)
+        rValues.resize(LocalSize,false);
+
+    noalias(rValues) = ZeroVector(LocalSize);
+
+    unsigned int index = 0;
+
+    for (unsigned int i = 0; i < NumNodes; i++) {
+        const array_1d<double,3>& r_acceleration = r_geometry[i].FastGetSolutionStepValue(ACCELERATION,Step);
+        for (unsigned int d = 0; d < Dim; d++)
+            rValues[index++] = r_acceleration[d];
+        rValues[index++] = 0.0; // skip pressure Dof
+    }
+}
+
+
+template< class TElementData >
+GeometryData::IntegrationMethod FluidElement<TElementData>::GetIntegrationMethod() const
+{
+    return GeometryData::GI_GAUSS_2;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Inquiry
+
+template< class TElementData >
+int FluidElement<TElementData>::Check(const ProcessInfo &rCurrentProcessInfo)
+{
+    // Generic geometry check
+    int out = Element::Check(rCurrentProcessInfo);
+
+    // Check variables used by TElementData
+    out = TElementData::Check(*this,rCurrentProcessInfo);
+
+    // Extra variables used in computing projections
+    KRATOS_CHECK_VARIABLE_KEY(ACCELERATION);
+
+    for(unsigned int i=0; i<NumNodes; ++i)
+    {
+        Node<3>& rNode = this->GetGeometry()[i];
+        KRATOS_CHECK_VARIABLE_IN_NODAL_DATA(ACCELERATION,rNode);
+
+        // Check that required dofs exist
+        KRATOS_CHECK_DOF_IN_NODE(VELOCITY_X,rNode);
+        KRATOS_CHECK_DOF_IN_NODE(VELOCITY_Y,rNode);
+        if (Dim == 3) KRATOS_CHECK_DOF_IN_NODE(VELOCITY_Z,rNode);
+        KRATOS_CHECK_DOF_IN_NODE(PRESSURE,rNode);
+    }
+
+    // If this is a 2D problem, check that nodes are in XY plane
+    if ( Dim == 2)
+    {
+        for (unsigned int i=0; i<NumNodes; ++i) {
+            if (this->GetGeometry()[i].Z() != 0.0)
+                KRATOS_ERROR << "Node " << this->GetGeometry()[i].Id() << "has non-zero Z coordinate." << std::endl;
+        }
+    }
+
+    return out;
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Input and output
+
+
+template< class TElementData >
+std::string FluidElement<TElementData>::Info() const
+{
+    std::stringstream buffer;
+    buffer << "FluidElement #" << Id();
+    return buffer.str();
+}
+
+
+template< class TElementData >
+void FluidElement<TElementData>::PrintInfo(std::ostream& rOStream) const
+{
+    rOStream << "FluidElement" << Dim << "D";
+}
+
+///////////////////////////////////////////////////////////////////////////////////////////////////
+// Protected functions
+///////////////////////////////////////////////////////////////////////////////////////////////////
+
+template <class TElementData>
+double FluidElement<TElementData>::Interpolate(typename TElementData::NodalScalarData& rValues,
```

Comment:
I've left it non-const because I understood you might use it to modify shape functions on the fly for the more complicated embedded formulations. If you don't need this, tell me and I'll gladly change it to const.